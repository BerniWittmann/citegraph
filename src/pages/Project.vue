<template>
  <div class="project">
    <v-alert v-if="!hasProject" type="error" prominent>
      {{ $t('project.not_available') }}
    </v-alert>

    <div v-if="hasProject">
      <h1>{{ project.name }}</h1>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project from '@/models/project'

@Component
export default class ProjectPage extends Vue {
  get project (): Project | undefined {
    return this.$store.getters['projects/activeProject']
  }

  get hasProject (): boolean {
    return this.$store.getters['projects/hasActiveProject']
  }
}
</script>
