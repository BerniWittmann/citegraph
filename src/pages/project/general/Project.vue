import { ProjectStates } from '@/models/project'
<template>
  <div class="project">
    <div class="d-flex mr-5 mt-4">
      <h1>{{ project.name }}</h1>
      <v-spacer/>
      <v-btn :to="{ name: 'projects.single.edit', params: { projectId: project.id } }">
        <v-icon>mdi-pencil</v-icon>
        {{ $t('project.edit.button_text') }}
      </v-btn>
    </div>
    <v-container v-if="hasNoDataYet">
      <v-row justify="center">
        <v-col :lg="3" class="d-flex justify-center flex-column text-center">
          <empty-icon class="align-self-center mb-6"></empty-icon>
          <h3 class="headline">{{ $t('project.single.no_data.description') }}</h3>
          <p>{{ $t('project.single.no_data.import_data') }}</p>
          <v-btn color="primary" rounded large :to="{ name: 'projects.single.import', params: { projectId: project.id } }">
            {{ $t('project.single.no_data.button') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-container>
    <v-container v-else>
      <v-row v-if="datasetInformation">
        <v-col>
          <v-card>
            <v-card-title>{{ $t('project.single.dataset.title') }}</v-card-title>
            <v-card-text>
              <v-container>
                <v-row>
                  <v-col v-for="(count, name) in datasetInformation" :key="name">
                    <v-card outlined class="d-flex flex-column justify-center text-center pa-5">
                      <p class="headline">{{count}}</p>
                      <p class="mb-0">{{ $t('project.single.dataset.' + name ) }}</p>
                    </v-card>
                  </v-col>
                </v-row>
              </v-container>
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
      <v-row>
        <v-col lg="3" md="6" cols="12" v-if="project.topAuthors">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_authors') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="author in project.topAuthors" :key="author.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'author',
                 entityId: author.id
                }}">
                  {{ author.firstName }} {{ author.lastName }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.topRecords">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_records') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="record in project.topRecords" :key="record.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'record',
                 entityId: record.id
                }}">
                  {{ record.title }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.countryRelevanceData">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.country_relevance') }}
            </v-card-title>
            <v-card-text>
              TODO Diagram here
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12">
          <v-row no-gutters>
            <v-col v-if="project.topKeywords" class="mb-6">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.top_10_keywords') }}
                </v-card-title>
                <v-card-text>
                  <v-chip-group multiple column>
                    <v-chip v-for="keyword in project.topKeywords" :key="keyword">{{ keyword }}</v-chip>
                  </v-chip-group>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
          <v-row no-gutters>
            <v-col v-if="project.counts">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.visualizations') }}
                </v-card-title>
                <v-card-text>
                  <div class="d-flex flex-column justify-center text-center pa-5" v-if="hasNoVisualizationsYet">
                    <p class="mb-0">{{ $t('project.single.overview.no_visualizations_yet') }}</p>
                    <v-btn color="primary" rounded>{{ $t('project.single.overview.add_visualization') }}</v-btn>
                  </div>
                  <v-card outlined class="d-flex flex-column justify-center text-center pa-5" v-else>
                    <p class="headline">{{ project.counts.visualizations }}</p>
                    <p class="mb-0">{{ $t('project.single.overview.visualizations') }}</p>
                  </v-card>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project, { ProjectStates } from '@/models/project'
import EmptyIcon from '@/assets/svgs/Empty.vue'

@Component({
  components: {
    EmptyIcon
  }
})
export default class ProjectPage extends Vue {
  get project (): Project {
    return this.$store.getters['projects/activeProject']
  }

  get datasetInformation (): Object | undefined {
    if (!this.project.counts) return undefined
    const data = { ...this.project.counts }
    delete data['visualizations']
    return data
  }

  get hasNoDataYet (): boolean {
    return this.project.state === ProjectStates.IMPORT
  }

  get hasNoVisualizationsYet (): boolean {
    return this.project.state === ProjectStates.EXPLORE
  }
}
</script>
