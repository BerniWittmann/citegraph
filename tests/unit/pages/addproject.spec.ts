import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'
import AddProjectPage from '@/pages/projects/AddProject.vue'
import Project from '@/models/project'

describe('pages/AddProject.vue', () => {
  const dispatch = jest.fn().mockReturnValue({
    then: (cb: Function) => cb(new Project({
      id: 42,
      name: 'My new Project'
    }))
  })
  const router = {
    push: jest.fn(),
    back: jest.fn()
  }
  const getWrapper = () => {
    return shallowMount(AddProjectPage, {
      i18n,
      mocks: {
        $store: {
          dispatch
        },
        $router: router,
        $t: (key: string) => key
      }
    })
  }
  it('renders the page', () => {
    const wrapper = getWrapper()
    // unset Rules to ensure Snapshot stability
    // @ts-ignore
    wrapper.vm.nameRules = []
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('it has a text field of the name', () => {
    it('has a text field', () => {
      const wrapper = getWrapper()
      expect(wrapper.contains('v-text-field-stub')).toBeTruthy()
    })
    it('the text needs to be entered', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      const rules = wrapper.vm.nameRules
      expect.assertions(rules.length)
      for (let rule of rules) {
        expect(rule('test')).toBeTruthy()
      }
    })
    it('the text is required', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      const rule = wrapper.vm.nameRules[0]
      expect(rule(undefined)).toEqual('projects.add.name_required')
      expect(rule(null)).toEqual('projects.add.name_required')
      expect(rule('')).toEqual('projects.add.name_required')
    })
    it('the text can not be an empty string', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      const rule = wrapper.vm.nameRules[1]
      expect(rule('    ')).toEqual('projects.add.name_not_empty')
      expect(rule('   test   ')).toBeTruthy()
      expect(rule(' ')).toEqual('projects.add.name_not_empty')
    })
  })

  it('has a cancel button', () => {
    const wrapper = getWrapper()
    const button = wrapper.findAll('v-btn-stub').at(0)
    expect(button.exists()).toBeTruthy()
    expect(button.text()).toEqual('projects.add.cancel')

    // @ts-ignore
    wrapper.vm.cancel()

    expect(router.back).toHaveBeenCalled()
  })

  describe('the project can be created', () => {
    it('has a submit button', () => {
      const wrapper = getWrapper()
      const button = wrapper.findAll('v-btn-stub').at(1)
      expect(button.exists()).toBeTruthy()
      expect(button.text()).toEqual('projects.add.submit')
    })
    it('the submit button is disabled if the form is not valid', () => {
      const wrapper = getWrapper()
      const button = wrapper.findAll('v-btn-stub').at(1)
      expect(button.exists()).toBeTruthy()
      expect(button.attributes('disabled')).toEqual('true')
      // @ts-ignore
      wrapper.vm.valid = true
      expect(button.attributes('disabled')).toBeUndefined()
    })
    it('does not create the project if form is not valid', (done) => {
      const wrapper = getWrapper()
      // @ts-ignore
      wrapper.vm.name = 'My new Project'
      // @ts-ignore
      wrapper.vm.valid = false

      expect.assertions(1)
      // @ts-ignore
      wrapper.vm.submit().then(() => {
        expect(dispatch).not.toHaveBeenCalled()
        done()
      })
    })
    it('creates the project on clicking the submit button', (done) => {
      const wrapper = getWrapper()
      // @ts-ignore
      wrapper.vm.name = 'My new Project'
      // @ts-ignore
      wrapper.vm.valid = true

      expect.assertions(2)
      // @ts-ignore
      wrapper.vm.submit().then(() => {
        expect(dispatch).toHaveBeenCalledWith('projects/createProject', new Project({
          id: undefined,
          name: 'My new Project'
        }))
        expect(router.push).toHaveBeenCalledWith({
          name: 'projects.single',
          params: {
            projectId: '42'
          }
        })
        done()
      })
    })
  })
})
