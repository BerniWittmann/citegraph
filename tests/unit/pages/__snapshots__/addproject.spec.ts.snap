// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/AddProject.vue renders the page 1`] = `
<div class="project-add">
  <v-container-stub tag="div">
    <v-row-stub tag="div">
      <h1 class="display-1 mt-5">projects.add.title</h1>
    </v-row-stub>
    <v-form-stub class="mt-12">
      <v-row-stub tag="div" justify="center">
        <v-col-stub lg="6" tag="div">
          <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" backgroundcolor="" hint="projects.add.name_hint" label="projects.add.name" loaderheight="2" autofocus="true" clearicon="$clear" solo="true" type="text" required=""></v-text-field-stub>
        </v-col-stub>
      </v-row-stub>
      <v-row-stub tag="div" justify="center" direction="row" class="mt-5">
        <v-col-stub lg="6" tag="div" class="d-flex flex-row">
          <v-btn-stub tag="button" activeclass="" text="true" type="button">projects.add.cancel</v-btn-stub>
          <v-spacer-stub></v-spacer-stub>
          <v-btn-stub color="success" tag="button" activeclass="" disabled="true" type="submit">projects.add.submit
          </v-btn-stub>
        </v-col-stub>
      </v-row-stub>
    </v-form-stub>
  </v-container-stub>
</div>
`;
